src/pairwise.jl

```diff
@@ -307,34 +314,130 @@ julia> pairwise(cor, eachcol(y), skipmissing=:pairwise)
-                            "a single set of variables (x === y)"))
-    end
-
+    check_vectors(x, y, skipmissing, symmetric)
     return _pairwise(Val(skipmissing), f, x, y, symmetric)
 end
 
-# cov(x) is faster than cov(x, x)
-_cov(x, y) = x === y ? cov(x) : cov(x, y)
+# Auxiliary functions for pairwise
+
+promoted_type(x) = mapreduce(eltype, promote_type, x, init=Union{})
+promoted_nmtype(x) = mapreduce(nonmissingtype ∘ eltype, promote_type, x, init=Union{})
+
+"""
+    handle_listwise(x, y)
+
+Remove missings in a listwise manner. Assumes `x` and `y` are vectors of iterables which
+have been validated via `check_vectors`.
+
+## Example
+```julia-repl
+julia> a = [6,7,8,9,10,missing];b = [4,5,6,7,missing,8];c = [2,3,4,missing,5,6];d = [1,2,3,4,5,6];
+
+julia> StatsBase.handle_listwise([a,b],[c,d])
+([[6, 7, 8], [4, 5, 6]], [[2, 3, 4], [1, 2, 3]])
+```
+"""
+function handle_listwise(x, y)
+    if !(missing isa promoted_type(x) || missing isa promoted_type(y))
+        return x, y
+    end
+
+    nminds = .!ismissing.(first(x))
+    @inbounds for xi in Iterators.drop(x, 1)
+        nminds .&= .!ismissing.(xi)
+    end
+    if x !== y
+        @inbounds for yj in y
+            nminds .&= .!ismissing.(yj)
+        end
+    end
+
+    # Computing integer indices once for all vectors is faster
+    nminds′ = findall(nminds)
+    # TODO: check whether wrapping views in a custom array type which asserts
+    # that entries cannot be `missing` (similar to `skipmissing`)
+    # could offer better performance
 
-pairwise!(::typeof(cov), dest::AbstractMatrix, x, y;
-          symmetric::Bool=false, skipmissing::Symbol=:none) =
-    pairwise!(_cov, dest, x, y, symmetric=symmetric, skipmissing=skipmissing)
+    x′ = [disallowmissing(view(xi, nminds′)) for xi in x]
+    if x === y
+        return x′, x′
+    else
+        y′ = [disallowmissing(view(yi, nminds′)) for yi in y]
+        return x′, y′
+    end
+end
 
-pairwise(::typeof(cov), x, y; symmetric::Bool=false, skipmissing::Symbol=:none) =
-    pairwise(_cov, x, y, symmetric=symmetric, skipmissing=skipmissing)
+"""
+    handle_pairwise(x::AbstractVector, y::AbstractVector;
+    scratch_fx::AbstractVector=similar(x, nonmissingtype(eltype(x))),
+    scratch_fy::AbstractVector=similar(y, nonmissingtype(eltype(y))))
 
-pairwise!(::typeof(cov), dest::AbstractMatrix, x;
-         symmetric::Bool=true, skipmissing::Symbol=:none) =
-    pairwise!(_cov, dest, x, x, symmetric=symmetric, skipmissing=skipmissing)
+Return a pair `(a,b)`, filtered copies of `(x,y)`, in which elements `x[i]` and
+`y[i]` are excluded if  `ismissing(x[i])||ismissing(y[i])`.
+"""
+function handle_pairwise(x::AbstractVector, y::AbstractVector;
+    scratch_fx::AbstractVector=similar(x, nonmissingtype(eltype(x))),
+    scratch_fy::AbstractVector=similar(y, nonmissingtype(eltype(y))))
+
+    axes(x) == axes(y) || throw(DimensionMismatch("x and y have inconsistent dimensions"))
+    lb = first(axes(x, 1))
+    j = lb - 1
```

Comment:
This seems a bit dangerous since `j` will be indexed into `scratch_fx` and `scratch_fy` but not `x` (or `y`) and initially `j` will be an invalid index (could be resolved be reordering `j += 1` in the loop).

Reply:
turns out I also needed to change the last line of the function which is now:

```
function handle_pairwise(x::AbstractVector, y::AbstractVector;
    scratch_fx::AbstractVector=similar(x, nonmissingtype(eltype(x))),
    scratch_fy::AbstractVector=similar(y, nonmissingtype(eltype(y))))

    axes(x) == axes(y) || throw(DimensionMismatch("x and y have inconsistent dimensions"))
    lb = first(axes(x, 1))
    j = lb
    @inbounds for i in eachindex(x)
        if !(ismissing(x[i]) || ismissing(y[i]))
            scratch_fx[j] = x[i]
            scratch_fy[j] = y[i]
            j += 1
        end
    end

    return view(scratch_fx, lb:(j-1)), view(scratch_fy, lb:(j-1))
end
```

Reply:
Fixed in commit 2119804